File: python/src/Game.py
import copy
import heapq
import sys

sys.path.insert(0, "../src")

from GenericBaseClass import *
from Solver import *
from WordChainDict import *

class Game(GenericBaseClass):
    OK = "ok"
    NOT_A_WORD = "not a word"
    NOT_ONE_STEP = "not one step away"
    DEAD_END = "no solution from this word"
    DUPLICATE = "word already played"

    def __init__(self, wordChainDict, solution):
        self.wordChainDict = wordChainDict
        self.solution = solution
        firstWord = solution.getFirstWord()
        self.solutionInProgress = Solution([firstWord], solution.target)

    def isSolved(self):
        return (self.solutionInProgress.isSolved() or
                self.solutionInProgress.getLastWord() == self.solution.getPenultimateWord())

    def playReplaceChar(self, index, c):
        # replace char at index in last word with 'c' 
        inWord = self.solutionInProgress.getLastWord()
        newWord = f"{inWord[:index]}{c}{inWord[index+1:]}"
        return self.playWord(newWord)

    def playRemoveChar(self, index): 
        #index is given as one of [0 .. len)
        inWord = self.solutionInProgress.getLastWord()
        newWord = f"{inWord[:index]}{inWord[index+1:]}"
        return self.playWord(newWord)
        
    def playInsertChar(self, index, c):
        # insert char at index in last word with 'c' 
        inWord = self.solutionInProgress.getLastWord()
        newWord = f"{inWord[:index]}{c}{inWord[index:]}"
        return self.playWord(newWord)

    def playWord(self, word):
        if not self.wordChainDict.isWord(word):
            return Game.NOT_A_WORD

        lastWordPlayed = self.solutionInProgress.getLastWord()

        if not word in self.wordChainDict.findNextWords(lastWordPlayed):
            return Game.NOT_ONE_STEP

        if self.solutionInProgress.isWordInSolution(word):
            return Game.DUPLICATE
            
        nextStep = self.solutionInProgress.numSteps() + 1
        if self.solution.getWordByStep(nextStep) == word:
            self.solutionInProgress.addWord(word)
            return Game.OK

        newSolver = Solver(self.wordChainDict, word, self.solution.getTarget())
        potentialNewSolution = newSolver.solveIt(solutionPrefix=self.solutionInProgress.getWordList())
        if potentialNewSolution.isSolved():
            self.solution = potentialNewSolution
            self.solutionInProgress.addWord(word)
            return Game.OK
        else:
            return Game.DEAD_END

    def showGame(self):
        playedWords = self.solutionInProgress.getWordList()
        solutionWords = self.solution.getWordList()
        resultStr = "";

        for i in range(0, len(solutionWords)):
            stepStr = "{:5s}".format("[{}]".format(i))
            if (i <= self.solutionInProgress.numSteps()):
                resultStr += "{}{}\n".format(stepStr, playedWords[i])
            else:
                resultStr += "{}{}\n".format(stepStr, self.showUnplayedWord(solutionWords[i], solutionWords[i-1]))
        return resultStr

    def showSolution(self):
        return str(self.solution.getWordList())


    # words not yet played should just use ***** and '!' to indicate shape of changes
    def showUnplayedWord(self, word, previousWord):
        unplayedWord = ""
        if len(word) == len(previousWord):
            for position, letter in enumerate(word):
                if word[position] == previousWord[position]:
                   unplayedWord += "*"
                else:
                    unplayedWord += "!"
        else:
            # add or remove a letter yielding n chars
            unplayedWord = "*****************"[0:len(word)]
                
        return unplayedWord 

    def showWordHint(self, word, previousWord):
        hint = ""
        if len(word) == len(previousWord):
            hint = self.showUnplayedWord(word, previousWord)
        elif len(word) < len(previousWord):
            # remove a letter numbered 1..len(previousWord)
            hint =  "123456789abcde"[0:len(previousWord)]
        else:
            # add a letter somewhere in previousi, including before and after: 
            for i in range (0, len(previousWord)):
                hint += str(i)
                hint += previousWord[i]
            hint += str(len(previousWord))

        return hint

File: resources/javascript/Game.js
class Game extends BaseLogger {
    static OK = "ok";
    static NOT_ONE_STEP = "Your word is not one step away from the last word.";
    static DEAD_END = "No solution from this word.";
    static DUPLICATE = "You already played that word.";

    constructor(dict, solution) {
        super();
        this.dict = dict;
        this.solution = solution;
        let firstWord = solution.getFirstWord();
        this.solutionInProgress = new Solution([firstWord], solution.target);
    }

    isSolved () {
        return this.solutionInProgress.isSolved();
    }

    playWord(word) {
        // Display will validate that the word is a valid word.

        if (this.solutionInProgress.isWordInSolution(word)) {
            return Game.DUPLICATE;
        }
    
        let lastWordPlayed = this.solutionInProgress.getLastWord();
        if (! this.dict.findNextWords(lastWordPlayed).has(word)) {
            return Game.NOT_ONE_STEP;
        }

        let nextStep = this.solutionInProgress.numSteps() + 1;
        if (this.solution.getWordByStep(nextStep) === word) {
            this.solutionInProgress.addWord(word);
            return Game.OK;
        }

        //let newSolver = new Solver(this.dict, word, this.solution.getTarget());
        //let potentialNewSolution = newSolver.solveIt(this.solutionInProgress.getWords());
        const potentialNewSolution = Solver.fastSolve(this.dict, word, this.solution.getTarget());

        if (potentialNewSolution.isSolved()) {
            this.solution = potentialNewSolution;
            this.solutionInProgress.addWord(word);
            return Game.OK;
        } else {
            return Game.DEAD_END;
        }
    }

    showGame() {
        let playedWords = this.solutionInProgress.getWords();
        let solutionWords = this.solution.getWords();
        let resultStr = ""; 

        let game = []
        for (let i = 0; i < solutionWords.length; i++) {
            if (i <= this.solutionInProgress.numSteps()) {
                game.push(playedWords[i])
            } else {
                game.push(this.showUnguessedWord(solutionWords[i], solutionWords[i-1]));
            }
        }

        return game;
    }

    showUnguessedWord(word, previousWord) {
        let unguessedWord = ""
        if (word.length === previousWord.length) {
            for (let i = 0; i < word.length; i++) {
                if (word[i] == previousWord[i]) {
                    unguessedWord += "*";
                } else {
                    unguessedWord += "!";
                }
            }
        } else {
            unguessedWord = "*".repeat(word.length);
        }

        return unguessedWord;
    }
}
